lib/msal-core/samples/react-sample-app/src/AuthProvider.js

```diff
@@ -0,0 +1,160 @@
+    GRAPH_REQUESTS
+} from "./auth-utils";
+
+// If you support IE, our recommendation is that you sign-in using Redirect APIs
+const useRedirectFlow = isIE();
+
+export default C =>
+    class AuthProvider extends Component {
+        constructor(props) {
+            super(props);
+
+            this.state = {
+                account: null,
+                error: null,
+                emailMessages: null,
+                graphProfile: null
+            };
+        }
+
+        async acquireToken(request, redirect) {
+            return msalApp.acquireTokenSilent(request).catch(error => {
+                // Call acquireTokenPopup (popup window) in case of acquireTokenSilent failure
+                // due to consent or interaction required ONLY
+                if (requiresInteraction(error.errorCode)) {
+                    return redirect
+                        ? msalApp.acquireTokenRedirect(request)
+                        : msalApp.acquireTokenPopup(request);
+                }
+            });
+        }
+
+        async onSignIn(redirect) {
+            if (redirect) {
+                return msalApp.loginRedirect(GRAPH_REQUESTS.LOGIN);
+            }
+
+            const loginResponse = await msalApp
+                .loginPopup(GRAPH_REQUESTS.LOGIN)
+                .catch(() => {
+                    this.setState({
+                        error: "Unable to acquire access token."
+                    });
+                });
+
+            if (loginResponse) {
+                this.setState({
+                    account: loginResponse.account,
+                    error: null
+                });
+
+                const tokenResponse = await this.acquireToken(
+                    GRAPH_REQUESTS.EMAIL
+                );
+
+                const graphProfile = await fetchMsGraph(
+                    GRAPH_ENDPOINTS.ME,
+                    tokenResponse.accessToken
+                );
+
+                this.setState({
+                    graphProfile
+                });
+            }
+        }
+
+        onSignOut() {
+            msalApp.logout();
+        }
+
+        async onRequestEmailToken() {
+            const tokenResponse = await this.acquireToken(
+                GRAPH_REQUESTS.EMAIL,
+                useRedirectFlow
+            ).catch(e => {
+                this.setState({
+                    error: "Unable to acquire access token for reading email."
+                });
+            });
+
+            if (tokenResponse) {
+                return this.readMail(tokenResponse.accessToken);
+            }
+        }
+
+        async readMail(accessToken) {
+            const emailMessages = await fetchMsGraph(
+                GRAPH_ENDPOINTS.MAIL,
+                accessToken
+            );
+
+            this.setState({
+                emailMessages,
+                error: null
+            });
+        }
+
+        async componentDidMount() {
+            msalApp.handleRedirectCallback(error => {
```

Comment:
does the token actually come back here?

Reply:
if we are doing nothing maybe just set a n00p with a comment 

would be nice if we knew why it mounted multiple times

Reply:
The token does come back here, but because of the refresh that happens, we lose the ability to do anything it, such as making API calls. And yeah, I'm going to look into why the refresh happens.

Reply:
Also, I think the type definitions are wrong for this method, as it says that the signature is `(response, error)`, when it's really `(error, response)` (as you would expect).

Reply:
I'll PR a fix for that separately.